torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,199 @@
+          Block* if_block = if_n.node()->blocks().at(block_index);
+          Block* other_if_block = if_n.node()->blocks().at(1 - block_index);
+          for (size_t i = 0; i < n->outputs().size(); ++i) {
+            Value* block_output = if_block->outputs().at(i);
+            if (!block_output->type()->cast<IntType>()) {
+              continue;
+            }
+            // Value must be in scope for both blocks
+            if (!if_n.node()->isDominatedBy(block_output->node())) {
+              continue;
+            }
+            // one constant value one not
+            auto other_const_value =
+                constant_as<int64_t>(other_if_block->outputs().at(i));
+            if (!other_const_value ||
+                block_output->node()->kind() == prim::Constant) {
+              continue;
+            }
+            const auto& other_block_refinements = block_index == 0
+                ? false_block_refinements
+                : true_block_refinements;
+            c10::optional<int64_t> maybe_refine =
+                tryFindRefinement(block_output);
+            if (!maybe_refine && other_block_refinements.count(block_output)) {
+              maybe_refine = other_block_refinements.at(block_output);
+            }
+            if (maybe_refine && *maybe_refine == *other_const_value) {
+              if_n.outputs().at(i)->replaceAllUsesWith(block_output);
+              changed_ = true;
+            }
+          }
+        }
+
+        joinIfRefinements(
+            n,
+            throwing_blocks_,
+            block_refinements,
+            true_block_refinements,
+            false_block_refinements,
+            info_);
+      }
+    }
+
+    // this is useful for things like if block outputs,
+    // where the output value node may not be defined in this block
+    // but we have refined its use as a block output
+    for (size_t i = 0; i < b->outputs().size(); ++i) {
+      Value* input_v = b->outputs().at(i);
+      if (!input_v->type()->cast<IntType>()) {
+        continue;
+      }
+
+      if (auto refine = tryFindRefinement(input_v)) {
+        WithInsertPoint guard(b);
+        auto refine_constant =
+            graph_->insertConstant(static_cast<int64_t>(*refine));
+        b->replaceOutput(i, refine_constant);
+        changed_ = true;
+      }
+    }
+
+    active_refinements_.pop_back();
+    return block_refinements;
+  };
+
+  c10::optional<int64_t> tryFindRefinement(Value* v) {
+    for (const auto& ref : active_refinements_) {
+      auto maybe_refinement = ref->find(v);
+      if (maybe_refinement != ref->end()) {
+        return maybe_refinement->second;
+      }
+    }
+    return c10::nullopt;
+  }
+
+  std::shared_ptr<Graph> graph_;
+  // A stack of active refinements, one for each block
+  std::vector<ListRefinement*> active_refinements_;
+  // A map from Boolean Value * -> associated refinements
+  std::unordered_map<Value*, BooleanRefinementMapping> info_;
```

Comment:
`info_` is too generic a name, how about `bool_values_to_refinement_map` for readability?